File: src/core/helm/HelmExecutionException.java
/*
 * Copyright (C) 2023 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.fullstack.helm.client;

import com.hedera.fullstack.base.api.util.StringUtils;

/**
 * Exception thrown when the execution of the Helm executable fails.
 */
public class HelmExecutionException extends RuntimeException {

    /**
     * the default message to use when no message is provided
     */
    public static final String DEFAULT_MESSAGE = "Execution of the Helm command failed with exit code: %d";

    /**
     * the non-zero system exit code returned by the Helm executable or the operating system
     */
    private final int exitCode;

    /**
     * the standard output of the Helm executable
     */
    private final String stdOut;

    /**
     * the standard error of the Helm executable
     */
    private final String stdErr;

    /**
     * Constructs a new exception instance with the specified exit code and a default message.
     *
     * @param exitCode the exit code returned by the Helm executable or the operating system.
     */
    public HelmExecutionException(final int exitCode) {
        this(exitCode, String.format(DEFAULT_MESSAGE, exitCode), StringUtils.EMPTY, StringUtils.EMPTY);
    }

    /**
     * Constructs a new exception instance with the specified exit code, standard output and standard error.
     *
     * @param exitCode the exit code returned by the Helm executable or the operating system.
     * @param stdOut the standard output of the Helm executable.
     * @param stdErr the standard error of the Helm executable.
     */
    public HelmExecutionException(final int exitCode, final String stdOut, final String stdErr) {
        this(exitCode, String.format(DEFAULT_MESSAGE, exitCode), stdOut, stdErr);
    }

    /**
     * Constructs a new exception instance with the specified exit code, message, stdOut, and stdErr.
     *
     * @param exitCode the exit code returned by the Helm executable or the operating system.
     * @param message the detail message (which is saved for later retrieval by the getMessage() method).
     * @param stdOut the standard output of the Helm executable.
     * @param stdErr the standard error of the Helm executable.
     */
    public HelmExecutionException(final int exitCode, final String message, final String stdOut, final String stdErr) {
        super(message);
        this.exitCode = exitCode;
        this.stdOut = stdOut;
        this.stdErr = stdErr;
    }

    /**
     * Constructs a new exception instance with the specified exit code and cause using the default message.
     *
     * @param exitCode the exit code returned by the Helm executable or the operating system.
     * @param cause    the cause (which is saved for later retrieval by the getCause() method). (A null value is
     *                 permitted, and indicates that the cause is nonexistent or unknown.)
     */
    public HelmExecutionException(final int exitCode, final Throwable cause) {
        this(exitCode, String.format(DEFAULT_MESSAGE, exitCode), cause);
    }

    /**
     * Constructs a new exception instance with the specified exit code, message and cause.
     *
     * @param exitCode the exit code returned by the Helm executable or the operating system.
     * @param message  the detail message (which is saved for later retrieval by the getMessage() method).
     * @param cause    the cause (which is saved for later retrieval by the getCause() method). (A null value is
     *                 permitted, and indicates that the cause is nonexistent or unknown.)
     */
    public HelmExecutionException(final int exitCode, final String message, final Throwable cause) {
        super(message, cause);
        this.exitCode = exitCode;
        this.stdOut = StringUtils.EMPTY;
        this.stdErr = StringUtils.EMPTY;
    }

    /**
     * Returns the exit code returned by the Helm executable or the operating system.
     *
     * @return the exit code returned by the Helm executable or the operating system.
     */
    public int getExitCode() {
        return exitCode;
    }

    /**
     * Returns the standard output of the Helm executable.
     *
     * @return the standard output of the Helm executable.
     */
    public String getStdOut() {
        return stdOut;
    }

    /**
     * Returns the standard error of the Helm executable.
     *
     * @return the standard error of the Helm executable.
     */
    public String getStdErr() {
        return stdErr;
    }

    @Override
    public String toString() {
        return "HelmExecutionException{" + "message="
                + getMessage() + ", exitCode="
                + getExitCode() + ", stdOut='"
                + getStdOut() + '\'' + ", stdErr='"
                + getStdErr() + '\'' + '}';
    }
}

File: src/integration/helm/helm-execution-exception.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Exception thrown when the execution of the Helm executable fails.
 */
export class HelmExecutionException extends Error {
  /**
   * The default message to use when no message is provided
   */
  private static readonly DEFAULT_MESSAGE = 'Execution of the Helm command failed with exit code: %d';

  /**
   * The non-zero system exit code returned by the Helm executable or the operating system
   */
  private readonly exitCode: number;

  /**
   * The standard output of the Helm executable
   */
  private readonly stdOut: string;

  /**
   * The standard error of the Helm executable
   */
  private readonly stdErr: string;

  /**
   * Constructs a new exception instance with the specified exit code and a default message.
   * @param exitCode The exit code returned by the Helm executable or the operating system
   */
  constructor(exitCode: number);
  /**
   * Constructs a new exception instance with the specified exit code, standard output and standard error.
   * @param exitCode The exit code returned by the Helm executable or the operating system
   * @param stdOut The standard output of the Helm executable
   * @param stdErr The standard error of the Helm executable
   */
  constructor(exitCode: number, stdOut: string, stdError: string);
  /**
   * Constructs a new exception instance with the specified exit code, message, stdOut, and stdErr.
   * @param exitCode The exit code returned by the Helm executable or the operating system
   * @param message The detail message
   * @param stdOut The standard output of the Helm executable
   * @param stdErr The standard error of the Helm executable
   */
  constructor(exitCode: number, message: string, stdOut: string, stdError: string);
  /**
   * Constructs a new exception instance with the specified exit code and cause using the default message.
   * @param exitCode The exit code returned by the Helm executable or the operating system
   * @param cause The cause
   */
  constructor(exitCode: number, cause: Error);
  /**
   * Constructs a new exception instance with the specified exit code, message and cause.
   * @param exitCode The exit code returned by the Helm executable or the operating system
   * @param message The detail message
   * @param cause The cause
   */
  constructor(exitCode: number, message: string, cause: Error);

  constructor(
    exitCode: number,
    messageOrStdOutOrCause?: string | Error,
    stdErrorOrCause?: string | Error,
    stdErrorParameter?: string,
  ) {
    let message: string;
    let cause: Error | undefined;
    let stdOut = '';
    let stdError = '';

    if (messageOrStdOutOrCause instanceof Error) {
      // Constructor with exitCode and cause
      message = HelmExecutionException.DEFAULT_MESSAGE.replace('%d', exitCode.toString());
      cause = messageOrStdOutOrCause;
    } else if (typeof messageOrStdOutOrCause === 'string') {
      if (stdErrorOrCause instanceof Error) {
        // Constructor with exitCode, message, and cause
        message = messageOrStdOutOrCause;
        cause = stdErrorOrCause;
      } else if (typeof stdErrorOrCause === 'string') {
        if (stdErrorParameter) {
          // Constructor with exitCode, message, stdOut, and stdErr
          message = messageOrStdOutOrCause;
          stdOut = stdErrorOrCause;
          stdError = stdErrorParameter;
        } else {
          // Constructor with exitCode, stdOut, and stdErr
          message = HelmExecutionException.DEFAULT_MESSAGE.replace('%d', exitCode.toString());
          stdOut = messageOrStdOutOrCause;
          stdError = stdErrorOrCause;
        }
      } else {
        // Constructor with just exitCode
        message = HelmExecutionException.DEFAULT_MESSAGE.replace('%d', exitCode.toString());
      }
    } else {
      // Constructor with just exitCode
      message = HelmExecutionException.DEFAULT_MESSAGE.replace('%d', exitCode.toString());
    }

    super(message);
    this.name = 'HelmExecutionException';
    this.exitCode = exitCode;
    this.stdOut = stdOut;
    this.stdErr = stdError;
    if (cause) {
      this.cause = cause;
    }
  }

  /**
   * Returns the exit code returned by the Helm executable or the operating system.
   * @returns The exit code returned by the Helm executable or the operating system
   */
  getExitCode(): number {
    return this.exitCode;
  }

  /**
   * Returns the standard output of the Helm executable.
   * @returns The standard output of the Helm executable
   */
  getStdOut(): string {
    return this.stdOut;
  }

  /**
   * Returns the standard error of the Helm executable.
   * @returns The standard error of the Helm executable
   */
  getStdErr(): string {
    return this.stdErr;
  }

  /**
   * Returns a string representation of the exception.
   * @returns A string representation of the exception
   */
  override toString(): string {
    return `HelmExecutionException{message=${this.message}, exitCode=${this.getExitCode()}, stdOut='${this.getStdOut()}', stdErr='${this.getStdErr()}'}`;
  }
}
